File: src/core/componentes.py
#-*- coding:utf-8-*-
from sys import exit

class Componentes(object):
    def __init__(self, configuracao):
        """
        configuracao -> Nome do Arquivo de configuração
        """
        self.items = {}             #NOTE: seria bom criar get e set? itens está escrito em inglês
        #NOTE: valores não é um bom nome...
        self.valores = {} #utilizado na busca de um nome a partir de um valor.
        try:
            linhas = open(configuracao, 'r').readlines()
        except:
            print "Erro durante a abertura do arquivo de configuração dos componentes."
            exit()

        cod = 1
        for linha in linhas:
            if linha[-1] == '\n': linha = linha[:-1]
            linha = linha.split(',')
            if len(linha) > 2:
                self.items[linha[0]] = cod, linha[1], linha[2]
                self.valores[cod] = linha[0]
            else:
                #TODO: tratar erro
                pass
            cod <<= 1

    def adicionar(self, item, valor):
        return self.items[item][0] | valor

    def adicionarVarios(self, itens, valor):
        if len(itens) == 1:
            return self.adicionar(itens[0], valor)
        else:
            return self.adicionar(itens[0], self.adicionarVarios(itens[1:], valor))

    def remover(self, item, valor):
        return self.items[item][0] ^ valor

    def checar(self, item, valor):
        #sei que não precisa de != 0
        return (self.items[item][0] & valor) != 0

    @staticmethod
    def juntar(v1, v2):
        return v1 | v2

    @staticmethod
    def separar(v1, v2):
        return v1 ^ v2

File: src/examples/aspirador-cpp/componentes.cpp
#include "componentes.h"
	
Componentes::Componentes(const char* configuracao) {
	carregar(configuracao);
}


void Componentes::carregar(const char* configuracao) {
	items.clear();
    ifstream in;
    int indice;
    in.open(configuracao);
    string aux;
    string sub;
    int val = 1;

	while (in >> aux) {
		items[aux.substr(0, aux.find(','))] = val;
		val <<= 1;
		//aux.erase(0, aux.find(',')+1);
		//aux.erase(0, aux.find(',')+1);
		//aux.erase(0, aux.find(',')+1);						
		
	}
	in.close();
}

Componentes::Componentes() {}

int Componentes::adicionar(string& item, int valor) {
    return items[item] | valor;
}
   
int Componentes::adicionarVarios(list<string> its, int valor) {
    if (its.size() == 1) { 
        return this->adicionar(its.front(), valor);
    }
    else {
    	string s = its.front();
    	its.pop_front();
        return adicionar(s, adicionarVarios(its, valor));
    }
}
    
int Componentes::remover(string item, int valor) {
    return items[item] ^ valor;
}

bool Componentes::checar(string item, int valor) { 
    //sei que não precisa de != 0
    return (items[item] & valor) != 0;
}

int Componentes::juntar(int v1, int v2) {
    return v1 | v2;
}

int Componentes::separar(int v1, int v2) {	
    return v1 ^ v2;
}

void Componentes::print() {
	map<string, int>::iterator mit;
	for (mit = items.begin(); mit != items.end(); ++mit)
		cout << (*mit).first << ' ' << (*mit).second << endl; 
}
Componentes::~Componentes() {
	//TODO
}
   
    
